docs/design/coreclr/botr/readytorun-format.md

```diff
@@ -35,7 +51,19 @@ The limitations of the current format are:
 
 # Structures
 
-The structures and accompanying constants are defined in the [readytorun.h](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/inc/readytorun.h) header file.
+The structures and accompanying constants are defined in the [readytorun.h]
+(https://github.com/dotnet/runtime/blob/master/src/coreclr/src/inc/readytorun.h) header file.
+Basically the entire R2R executable image is addressed through the READYTORUN_HEADER singleton
+pointed to by the well-known export RTR_HEADER in the export section of the native executable
+envelope.
+
+For single-file R2R executables, there's just one header representing all image sections.
```

Comment:
Seems a bit confusing, as far as I can tell there's only one header no matter if the file is composite or not.

Reply:
@mikedn - I must admit I'm somewhat unsure what you're trying to say: it seems to me you're either trying to say that CoreCLR runtime and Crossgen1/2 always only generate one header - that is correct, this is a feature proposal that requires adding the support for per-assembly sub-headers - or that the proposal is poorly written in the sense of insufficiently or inconsistently referring to header levels. Could you please clarify?

Based on how I'm imagining the design and based on JanK's comments the version I'm about to publish as the 3rd commit on this review distinguishes three different cases:

1) Legacy case - Windows MSIL PE with attached R2R data (single-assembly only);

2) Modern single-assembly R2R - initially Windows native PE, ultimately OS-native executable format, single R2R header locatable via well-known export;

3) Modern multiple-assembly composite / single-exe R2R - initially Windows native PE, ultimately OS-native executable format, master R2R header locatable via well-known export plus assembly table pointing to per assembly subheaders with assembly-level sections (e.g. AVAILABLE TYPES).

Reply:
> or that the proposal is poorly written in the sense of insufficiently or inconsistently referring to header levels. Could you please clarify?

It's a question of phrasing. Don't both single and composite files have a single `READYTORUN_HEADER` and the difference between them is the presence of a `READYTORUN_SECTION_ASSEMBLIES` section?

Reply:
That is true but the main purpose of the ASSEMBLIES section is to encode a series of per assembly headers (of the type <code>READYTORUN_CORE_HEADER</code>) that have a similar structure and purpose as the main R2R header - to encode a set of sections specific to a particular assembly. I'll take another look to see how to make it more obvious; I have also slightly reformulated this part in 3rd commit based on JanK's comments.

Reply:
Thanks, I think simply specifying the header name would be sufficient. Something like:

> For single-file R2R executables, there's just one `READYTORUN_CORE_HEADER` header representing all image sections.

In the composite case the text already says:

> Each `READYTORUN_SECTION_ASSEMBLIES_ENTRY` record points  to a `READYTORUN_CORE_HEADER` variable-length structure representing sections specific to the particular assembly.

so it should be clear enough.

Reply:
I have noticed that I missed updating the description of the ASSEMBLIES section. This should be fixed now as of 4th commit, I'll certainly appreciate if you could take another look to make sure the description is fully consistent.